### PR TITLE
sensors: test: Add attribute support to sensor emulator backend

### PR DESCRIPTION
**Pull request by yperess**
* Fix BMI160 to align with data sheet
* Add BMI160 emulator backend so the data collection tests can be done by the generic test
* Add attribute 'get' support to BMI160 and emulator backend
* Write custom tests for bus failures on the BMI160

### CODE REVIEW

File: drivers/i2c/i2c_emul.c
Comment (MaureenHelm) [inline drivers/i2c/i2c_emul.c:101]: This change looks unrelated to attributes?

File: drivers/i2c/i2c_emul.c
Comment (MaureenHelm) [inline drivers/i2c/i2c_emul.c:131]: This change is definitely not related to attributes

File: drivers/sensor/akm09918c/akm09918c_emul.c
Comment (MaureenHelm) [inline drivers/sensor/akm09918c/akm09918c_emul.c:167]: More unrelated clang-format changes

File: include/zephyr/drivers/i2c_emul.h
Comment (MaureenHelm) [inline include/zephyr/drivers/i2c_emul.h:50]: I2C changes should be in their own commit

File: drivers/sensor/bmi160/bmi160.c
Comment (MaureenHelm) [inline drivers/sensor/bmi160/bmi160.c:727]: New feature isn't related to the commit message `bmi160: Make changes to align driver to datasheet`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_emul.c
Comment (yperess) [reply drivers/i2c/i2c_emul.c]: Sorry, it's a part of the tests we're using in a related change downstream. I moved one such test upstream to use this as an example (more to follow).

File: drivers/sensor/akm09918c/akm09918c_emul.c
Comment (yperess) [reply drivers/sensor/akm09918c/akm09918c_emul.c:167]: Not unrelated `value` is now a pointer. Adding `*` required a format change due to line length.

File: drivers/sensor/bmi160/emul_bmi160.h
Comment (alevkoy) [inline drivers/sensor/bmi160/emul_bmi160.h:17]: Why are these defined in the header file?

Comment (yperess) [reply drivers/sensor/bmi160/emul_bmi160.h:17]: They're currently used in the bmi160 test suite, but I figured that these are fairly common functions for testing that would be used when verifying bus traffic. Where would you have them instead?

Comment (alevkoy) [reply drivers/sensor/bmi160/emul_bmi160.h:17]: I figured they should be defined in the .c file with the other bmi160 functions. If you envision using them in tests that would preclude linking in the .c file, I might recommend having some kind of common/utility .c file that could be linked in, but they're OK here too. You might want to leave a comment about where they're expected to be used.

**Review by alevkoy — DISMISSED** [review]

File: include/zephyr/drivers/emul_sensor.h
Comment (yperess) [reply include/zephyr/drivers/emul_sensor.h:36]: Attributes currently use the old representation of sensor values which are a pain point for many drivers. I purposefully don't want to add more APIs that use `sensor_value` as I believe it will be deprecated in the near future. Reading data will be using the mapping we've introduced with the decoder of channel -> data type. Attributes will likely follow suit. As such, I think it's best to have explicit functions which can provide these better data type bindings to the attributes. It's possible that once we land on the right attribute API, we'll have to collapse these into 1 function, but for now I would prefer not to keep building on the attribute API.

**Review by alevkoy — DISMISSED** [review]

File: include/zephyr/drivers/emul_sensor.h
Comment (tristan-google) [reply include/zephyr/drivers/emul_sensor.h:36]: I'm perfectly happy to dump `sensor_value` (we should be using Q31 exclusively in the backend API) or making it a generic pointer to allow for flexibility, but I disagree with creating separate API functions for each kind of attribute. Having a single one (1) reduces code duplication, (2) avoids needing to keep adding more functions when somebody wants the generic test to cover a new attribute, and (3) makes it more awkward for the test to discover supported attributes --with this solution the test would need to test-call each attribute setter function to discover capabilities (and also have a switch block in the test body for fanning out each type of attribute to the appropriate setter)

Comment (yperess) [reply include/zephyr/drivers/emul_sensor.h:36]: Ah, I see. I think this goes beyond just q31_t. I'm getting at the fact that some channel/attribute combinations are bools, some are 1D `q31_t` some are 3D `q31_t` some might be be counters like `uint64_t`. Until we're ready to tackle the attribute API, I don't want to keep building on top of it. Lets assume there's a unified backend API that always accepted `q31_t *values, size_t value_count` as arguments. If an attribute requires a `bool` how would you represent it?

I think it's easier to have separate functions that have the explicit value counts and types as part of the signature. Then when we clean up the attribute API it'll be easy to merge them if/when needed.

Comment (yperess) [reply include/zephyr/drivers/emul_sensor.h]: Done

**Review by alevkoy — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]
Seems like lots of C files but with .hpp/.cpp suffixes, not sure if that's what you looking for? I don't think Zephyr has any C++ directly in the tree for things like this does it? My -1 here is for this alone. I don't have a strong opinion on the rest.

More generally though...

Not entirely sure I fully grok the benefit of these tests without HIL to be honest either, were you able to stop/find some legitimate bugs?

Without the physical device around doing a HIL test against it, all the fun quirks and things are likely to fail to pop up. Having to mock out an entire register map to validate that the driver does indeed poke some registers correctly looks like a *lot* of work for questionable benefit to me.

What would it take instead to start having HIL testing with sensors is a question I've been asking myself, given the large number of drivers in the tree, the need to migrate them, and more.

I could very well be alone on this though.

File: tests/drivers/sensor/bmi160/src/i2c.cpp
Comment (yperess) [inline tests/drivers/sensor/bmi160/src/i2c.cpp:30]: @teburd this is the reason for C++. I can use a lambda here which makes the test much easier to read since the mock is inside the test function.

**Comment by yperess** [comment]
> Seems like lots of C files but with .hpp/.cpp suffixes, not sure if that's what you looking for? I don't think Zephyr has any C++ directly in the tree for things like this does it? My -1 here is for this alone. I don't have a strong opinion on the rest.

Added  a comment pointing at why C++. There are more tests coming and having it greatly simplifies the tests. It's even better downstream with Pigweed which I hope to one day have as a module.

> More generally though...
> 
> Not entirely sure I fully grok the benefit of these tests without HIL to be honest either, were you able to stop/find some legitimate bugs?
> 
> Without the physical device around doing a HIL test against it, all the fun quirks and things are likely to fail to pop up. Having to mock out an entire register map to validate that the driver does indeed poke some registers correctly looks like a _lot_ of work for questionable benefit to me.
> 
> What would it take instead to start having HIL testing with sensors is a question I've been asking myself, given the large number of drivers in the tree, the need to migrate them, and more.
> 
> I could very well be alone on this though.

I get what you're saying, but for example all the bugs included in this PR were found through emulation. It's also much more reproducible and runs much faster and cheaper. I can't speak for everyone, but at Google we rely very heavily on these emulators for end to end testing.

File: tests/drivers/sensor/bmi160/src/i2c.cpp
Comment (teburd) [reply tests/drivers/sensor/bmi160/src/i2c.cpp:30]: https://docs.zephyrproject.org/latest/develop/languages/cpp/index.html#enabling-c-support

The docs promise C++ support in application code. I don't see any notes about testing code. I generally find C++ incredibly difficult to read but I'm not the sole developer here, maybe writing tests in C++ would be ok, but thats something that likely needs to be brought up in the arch wg. I guess the next question would be what C++ standard? Seemingly C++98 is the only thing guaranteed and I don't believe lambdas are in 98.

Comment (yperess) [reply tests/drivers/sensor/bmi160/src/i2c.cpp:30]: I can't find it in the meeting notes (I guess it wasn't an important comment to write down), but I believe we had this discussion a long time ago in the TSC where we said that tests are fine in C++. This came up when I first added CHRE. I know we don't explicitly ban them:
image

There are more tests that we can enable via C++ class fixtures later along with the emulators. I personally feel like it gives us a lot of power to test things more easily and quickly.

Comment (henrikbrixandersen) [reply tests/drivers/sensor/bmi160/src/i2c.cpp:30]: I am with @teburd on this one. I do not recall having such a discussion in the TSC and I certainly do not support starting to use C++ in what is otherwise a C project with support for applications written in C++. Sorry.

**Comment by henrikbrixandersen** [comment]
> Without the physical device around doing a HIL test against it, all the fun quirks and things are likely to fail to pop up. Having to mock out an entire register map to validate that the driver does indeed poke some registers correctly looks like a lot of work for questionable benefit to me.
> 
> What would it take instead to start having HIL testing with sensors is a question I've been asking myself, given the large number of drivers in the tree, the need to migrate them, and more.
> 
> I could very well be alone on this though.

I 100% agree. This is close to what was being discussed in the TSC F2F in Prague earlier this year, where I raised this exact concern.

**Comment by yperess** [comment]
Removed C++

**Review by teburd — DISMISSED** [review]

**Review by alevkoy — DISMISSED** [review]

**Comment by yperess** [comment]
@henrikbrixandersen @MaureenHelm any chance to get another review pass before the end of the year? This is a blocker for us to use upstream drivers

**Review by tristan-google — DISMISSED** [review]

File: drivers/sensor/bmi160/bmi160.c
Comment (MaureenHelm) [inline drivers/sensor/bmi160/bmi160.c:254]: The commit message drops off mid-sentence:
```
There was also some extra conversions from SI->register->SI when
setting the range, this was causing the
```

File: tests/drivers/sensor/accel/src/main.c
Comment (MaureenHelm) [inline tests/drivers/sensor/accel/src/main.c]: Is this change in the wrong commit? It doesn't seem related to `bmi160: Make changes to align driver to datasheet`

**Comment by yperess** [comment]
The push above is just a rebase so that the `Compare` button with @MaureenHelm 's requested changes doesn't show the rebase diff. Changes coming in a few minutes.

File: tests/drivers/sensor/accel/src/main.c
Comment (yperess) [reply tests/drivers/sensor/accel/src/main.c]: This is correct, the extra tests here add coverage to the code that was fixed.

File: drivers/sensor/bmi160/bmi160.c
Comment (yperess) [reply drivers/sensor/bmi160/bmi160.c:254]: Done

File: drivers/sensor/bmi160/emul_bmi160.c
Comment (yperess) [reply drivers/sensor/bmi160/emul_bmi160.c:552]: I expanded them to show how I got them and documented.

**Review by alevkoy — APPROVED** [review]

**Review by tristan-google — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
